### PR TITLE
Iron

### LINKED ISSUE #93: Add iron to allow more advanced recipes
Since copper is already implemented in the game, I'd suggest adding iron in order to allow more advanced recipes. 

When it comes to the world gen, I'd suggest an `F` for the symbol (after the chemical element `Fe`) and choose either an orange color (then change copper to blue, in order to stay as close as possible to real world) or blue to stay close to what most games use those days.

We could add iron veins close to the mountains (and/or make terrains like suggested in #18). That would mean that we could get iron/copper in the wild (like it is now for copper) then have those terrains/veins to farm more for later use.

Finally, this would give options for the potential `robot chassis` suggested in #35 and also add/change other recipe (like the `drill` for example). Here's a few recipe suggestions to start:

- 1 iron ore -> 2 iron plates (1:2 or 1:1)
- 2 iron plates -> 1 iron gear (that would mean to rename the current `Gear` to this or differentiate `Wooden Gear` and `Iron Gear`)
- 2 iron plates + 1 iron gear + 2 copper pipes -> 1 robot chassis
- 1 robot chassis + 4 iron plates + 8 copper wires -> 1 robot

Those are obviously open for discussion.

### LINKED ISSUE #341: Compass device
In some challenges it is undesirable to be able to turn to cardinal directions.

For example in some algorithmic challenges the robot should only turn `left` or `right`, to teach thinking relative to the robot.

The chess horse challenge sets the robot direction to `[2,1]` so it moves like the knight piece. Turning north cheeses the challenge. :sweat_smile: 

## Solution
Create a compass device, some cardinal orientation ability which will be required for cardinal directions.

The recipe could be something really simple like:
```
1 water ─┬───── compass
1 copper ┘
   wire
```

## Alternatives

We could limit abilities directly in challenges, but it would still require creating some compass ability.

---

- credit for the idea goes to @byorgey in #339

### LINKED ISSUE #342: Suggest required entities
Currently, the exception thrown when the robot is missing some capability looks like this:
```
│ ───────────────────Log────────────────── │
│                                          │
│ missing device(s) needed to execute      │
│   command 'blocked'                      │
└──────────────────────────────────────────┘
```
Some commands are obvious, like `drill` requiring `drill`, others like `blocked` requiring `scanner` should be probably mentioned in the exception message.

In my mind there are three cases:
1. The player knows the necessary entity but the base/robot does not have it.
2. The player does not know the required entity.
3. There is no entity that the player could obtain to gain the capability (e.g. the `CGod` capability).

Giving a nice message to each of those cases could be slightly harder, but would IMO help beginners a lot.

### PR DESCRIPTION
**Pull request by xsebek**
- add iron ore, iron mine and iron vein (closes #93)
  - split gear into iron/wooden gear
  - add metal drill
  - add faster recipes with the metal drill
- add compass (closes #341)
- handle multiple entities providing the same capability
  - try to find if the robot has at least one entity providing the capability
  - when no entity could provide the capability rejects it too
- list required devices in the `Incapable` error (closes #342)

### CODE REVIEW

File: data/entities.yaml
Comment (byorgey) [inline data/entities.yaml:146]: Seems like this description got copied from `iron ore` and never edited?

File: data/entities.yaml
Comment (byorgey) [inline data/entities.yaml:159]:
```suggestion
    attr: iron
```

File: data/entities.yaml
Comment (byorgey) [inline data/entities.yaml:160]:
```suggestion
    char: 'F'
```

File: data/entities.yaml
Comment (byorgey) [inline data/entities.yaml:336]:
```suggestion
    attr: iron
```

File: data/entities.yaml
Comment (byorgey) [inline data/entities.yaml:176]: So one idea I had was to get rid of the distinction between `copper vein`/`iron vein` and `copper mine`/`iron mine` and just have `vein` and `mine`.  Given #372, the idea is that drilling in a mine would give you copper with some probability and iron with another (smaller) probability. We can also easily do things like a wooden drill might get you copper but never iron, whereas an iron drill could get you both, etc. --- we just have to specify the right recipes. While this scenario of getting everything from a single mine may not be very accurate in the "real world", I like it for a few reasons:

- It keeps things simpler in terms of the number of different entities in the world.
- It leaves the door open for adding other mine-able things (gold? diamonds? etc.) with much lower probabilities, without having to add yet more separate types of veins and mines.
- With multiple different types of mines, the difficulty for a player lies in locating all the different types and building separate drilling operations at each. With a single mine, the difficulty for the player lies in building a drilling and processing/sorting operation.  The latter sounds much more fun and interesting from a programming perspective than the former.

What do you think?

File: src/Swarm/Game/Step.hs
Comment (byorgey) [inline src/Swarm/Game/Step.hs:886]: This should not be a `Fatal` error.  The  `Fatal` error from before was for the situation that looking up `drill` in *the entity map* failed, i.e. the system as a whole did not know about the `drill` entity at all.  But the new code is just checking what kind of drill the robot has installed.  It could certainly happen that the robot is trying to execute a `drill` command even though it has no drill installed.  That is a runtime error but not a `Fatal` one.

**Review by byorgey — COMMENTED** [review]
Some initial comments.  I haven't finished looking carefully through the changes regarding how we check for required devices etc.

File: data/entities.yaml
Comment (xsebek) [reply data/entities.yaml:176]: I actually like both ideas - maybe we can split it into different scenarios :thinking: 

It does not require any extra logic in Swarm so the only overhead is extra recipes in our main `data/recipes.yaml`.
I think moving it to some *adventure mode* scenario would both clean up and leave our options open.

My only concern is how to order the changes - I will have to check if scenarios support all I need now or if it is easier to move it over later. :slightly_smiling_face:

File: src/Swarm/Game/Step.hs
Comment (xsebek) [reply src/Swarm/Game/Step.hs:886]: I suspected as much, but what example would escape our capability checking ahead of drilling?

**Comment by xsebek** [comment]
Thanks for taking a look @byorgey, the copy-pasting is definitely a problem :sweat_smile: 

I spent most type debugging the capability checking, so now that it (hopefully?) works I will come back and streamline it.

File: data/entities.yaml
Comment (xsebek) [reply data/entities.yaml:176]: Alternatively, we could parametrize entities by their underlying material to cut down on boilerplate.

I am not sure how it would work, but maybe it will become more evident once we add gold and other late-game stuff.

Comment (byorgey) [reply data/entities.yaml:176]: You said you like both ideas.  What is it you like about the idea of having separate `copper mine`, `iron mine`, etc.?

I don't mind supporting both, and I am not so much concerned about boilerplate as I am about the gameplay experience.  I don't mind having extra entities and recipes in `recipes.yaml` etc.  So we could have both `mine` that works as I described with weighted random results as well as specific `iron mine`, `copper mine`, etc.  The biggest issues then would be (1) which version do you get in the "default" mode?  Even if we support both we still have to decide which is the default game.  And (2) in order to support both we have to allow multiple "world functions" and provide a mechanism for scenarios to specify which world function they will use.  I think this was discussed in an issue somewhere but I can't find it at the moment.

File: src/Swarm/Game/Step.hs
Comment (byorgey) [reply src/Swarm/Game/Step.hs:886]: Hmm, good question.  I guess you're right, if this happens it means something has gone wrong with capability checking, so perhaps that should actually be a `Fatal` error.

**Comment by byorgey** [comment]
By the way, I very much like all the new recipes, the way metal drill does things faster than a regular drill, etc.

File: data/entities.yaml
Comment (xsebek) [reply data/entities.yaml]: Fixed :+1:

**Comment by xsebek** [comment]
That Restyled action is a bit annoying since I by mistake pushed to my repo which confused Restyled and it wants to merge to `main`. :facepalm:

File: data/entities.yaml
Comment (xsebek) [reply data/entities.yaml:176]: I like that it is visual, easier to debug and requires sending robots a bit further. :slightly_smiling_face: 

We could actually do with only mines and have them randomly drilled from `mountain`. That would address the number of entities in the world and make it simpler to add new mines, but it would be even harder to establish mining operations.

Comment (byorgey) [reply data/entities.yaml:176]: Hmm, I like the idea of mines appearing from a drilled mountain with a certain probability (in all cases, whether there is only one kind of mine or multiple).  It requires more exploration.  It also feels weird to me that right now the mountains are different colors depending on whether there is a vein --- not that Swarm needs to reflect the real world necessarily, but if you have to drill for something, you should not be able to tell just by looking at the surface what you will find when you drill.

File: bench/Benchmark.hs
Comment (byorgey) [inline bench/Benchmark.hs:63]: Let's change this to `turn right; move; turn right; move; ...` i.e. with the same number of turns and moves as before.  Otherwise the benchmark won't really be comparable to before, since the number of straight-line instructions the robot executes vs the number of recursive calls to `forever` can definitely make a performance difference.

File: data/recipes.yaml
Comment (byorgey) [inline data/recipes.yaml:228]: I like it.

File: src/Swarm/Game/Step.hs
Comment (byorgey) [inline src/Swarm/Game/Step.hs:388]: I am not sure about this.  Should we really need the `orient` capability just to *evaluate* a cardinal direction constant?

Comment (xsebek) [reply src/Swarm/Game/Step.hs:388]: Well, `base` would escape the check if I did not do it, but it still seems wrong:
```Haskell
> turn north -- compass is needed (but the error does not mention it)
> let d = north in build {turn d} -- once base has compass, nobody else needs it
```
I probably just overthought it and checking it when turning and drilling would work just fine.

- related #201 for the case when a robot is sent in some direction and it turns out it is cardinal at runtime (I should write a test for it...)
- related error message improvement #342

File: bench/Benchmark.hs
Comment (xsebek) [reply bench/Benchmark.hs:63]: Indeed it does seem to be 20% slower, I should have checked :+1:

File: data/entities.yaml
Comment (xsebek) [reply data/entities.yaml:156]: I think I found a good balance:
- mountains can be drilled with an 8/1/1 probability to tunnel/copper mine/iron mine
- in scenarios, we can use the deterministic vein ➔ mine custom recipes and entities

File: bench/Benchmark.hs
Comment (xsebek) [reply bench/Benchmark.hs:63]: Fixed :heavy_check_mark:

File: src/Swarm/Game/Step.hs
Comment (xsebek) [reply src/Swarm/Game/Step.hs]: Reverted and moved the check to commands using the direction.

File: src/Swarm/Language/Syntax.hs
Comment (xsebek) [inline src/Swarm/Language/Syntax.hs:123]: I think I might have at some point advocated for `down` being a cardinal direction, but now I need it to be allowed without a compass :sweat_smile: I do not think it's worth coming up with another name for the set of directions that are cardinal but not `down`. Writing `/= down` everywhere is also an option, but not a good one. :shrug:

**Comment by xsebek** [comment]
Hey @byorgey, could you please take a look and test if this works?

I made a testing playground with the old `copper vein`, `iron vein` and both drills and `compass` in inventory:
```bash
cabal run swarm:swarm -- --scenario 03Challenges/02-drill_test --cheat
```
Screenshot from 2022-06-11 02-22-12

I did not get around to automating the tests yet, but I would be happy to let someone else test my fragile code if they felt like it :wink:

File: src/Swarm/Game/Step.hs
Comment (xsebek) [reply src/Swarm/Game/Step.hs:904]: Given that even build inside build checks ahaed, I am going to leave it as Fatal and see if it comes up.

```
> build {r <- build {move; drill down; log ""}; log ""};
<r3> : robot
> salvage
```
```
[busy_williamson] build: You do not have required devices:
  - metal drill or drill
  - logger
```

File: data/entities.yaml
Comment (byorgey) [reply data/entities.yaml:156]: Ah, OK, I like that.

**Comment by byorgey** [comment]
The test scenario seems to work.  However, when I started creative mode and tried sending a robot to drill some mountains, it did not work.  Maybe it is failing due to having no `drill` installed even though we are in creative mode?

File: src/Swarm/Game/Step.hs
Comment (byorgey) [inline src/Swarm/Game/Step.hs:1427]: Took me a little while to figure out what this is for.  Maybe add a comment, something like: ignoreOK is given two lists of devices which all provide a given capability, a list of OK devices and a list of the other devices. If *at least one* device providing a capability is OK (i.e. the first list is not empty), then we can ignore all the alternative devices that are not installed.

File: src/Swarm/Util/Yaml.hs
Comment (byorgey) [inline src/Swarm/Util/Yaml.hs:65]: I would call this `getE`.  `fromE` sounds like you're converting `e` to some other type.

File: src/Swarm/Game/Step.hs
Comment (byorgey) [inline src/Swarm/Game/Step.hs:1444]: To make sure I'm understanding correctly, if this "impossible task" error is triggered, it means that there do not exist any devices to provide the required capability, right?  As in, we have some capability but no actual devices yet in `entities.yaml` which provide it?

File: src/Swarm/Language/Syntax.hs
Comment (byorgey) [reply src/Swarm/Language/Syntax.hs:123]: Yeah, making `down` a relative direction makes sense, given the existence of `compass`.  You should not need a `compass` to find the ground. :laughing:

**Review by byorgey — COMMENTED** [review]
Everything looks good to me, but there's still something I don't understand going on with drilling in creative mode.

File: src/Swarm/Game/Step.hs
Comment (byorgey) [reply src/Swarm/Game/Step.hs:1444]: Hmm, I actually got this error while playing in classic mode and it's very mysterious.  How hard would it be to output the name of the capability for which there is no device, with a message apologizing and a link to the issue tracker?

Comment (xsebek) [reply src/Swarm/Game/Step.hs:1444]: I expected this to only come up for `CGod` tier tasks, like `create "life"`, where the message would make sense.

Maybe I could check if it is `CGod` and otherwise link to the issue. Something like:
```Haskell
[ singularSubjectVerb subject "do", "not have", indefiniteQ capName
, "capability, because no device provides it."
, "\nSee https://github.com/swarm-game/swarm/issues/26."
]
```

However, I feel like it should be `Incapable` exceptions job to figure this out. :smirk:

File: src/Swarm/Util/Yaml.hs
Comment (xsebek) [reply src/Swarm/Util/Yaml.hs:65]: I was considering `askE`, to make it look like `ReaderT`, but `getE` sounds good too.

Comment (byorgey) [reply src/Swarm/Util/Yaml.hs:65]: Yeah, `askE` is probably better!

Comment (xsebek) [reply src/Swarm/Util/Yaml.hs]: Too late, I used your suggestion :laughing:

Comment (byorgey) [reply src/Swarm/Util/Yaml.hs]: Hah! Either one is fine. 😄

File: src/Swarm/Game/Step.hs
Comment (xsebek) [reply src/Swarm/Game/Step.hs]: Some of it was common, but building and reprogramming can give a more specialised error message, so I keep some duplicity code. Oh well, at least I get to close #342 :innocent:

File: src/Swarm/Game/Step.hs
Comment (xsebek) [reply src/Swarm/Game/Step.hs:1433]: Sure, I added some comments :+1:

File: data/scenarios/03Challenges/02-drill_test.yaml
Comment (xsebek) [reply data/scenarios/03Challenges/02-drill_test.yaml:72]: I suggest we solve it later, I made an Issue so we don't forget:
- #387

**Comment by xsebek** [comment]
> Everything looks good to me, but there's still something I don't understand going on with drilling in creative mode.

@byorgey it was actually more than drilling, by mistake I did not install any required entities if they were not present in inventory.

The commit [9e24293](https://github.com/swarm-game/swarm/pull/373/commits/9e2429394850d9f51a9b857abd46924798512663) should fix that :slightly_smiling_face:

File: data/scenarios/03Challenges/02-drill_test.yaml
Comment (byorgey) [reply data/scenarios/03Challenges/02-drill_test.yaml:72]: 👍

File: src/Swarm/Game/Exception.hs
Comment (byorgey) [inline src/Swarm/Game/Exception.hs:128]: I think this error message could still be improved.  I don't think it is going to make much sense if a player stumbles across it.  *e.g.* imagine that they type `as` accidentally and notice that it has a type, then they wonder what it does so they try it out. They are going to wonder *why* the task is impossible. Somehow we should convey that this is a command that is not available during normal play.  We can definitely make it funny though.  Some ideas:
- "This command is not available to mere mortals" / "This command is forbidden"
- "Thou shalt not utter such blasphemy"
- If we wanted to be more informative and less funny, "This command is only available to system robots and scenario definitions".

**Review by byorgey — APPROVED** [review]

**Comment by byorgey** [comment]
This is looking great! :rocket:

File: src/Swarm/Game/Step.hs
Comment (byorgey) [reply src/Swarm/Game/Step.hs]: Also, just an update: the reason I got this error while playing in classic mode was actually due to #389 .  So that was making it more mysterious than necessary.

**Comment by byorgey** [comment]
But up to you whether we want to go ahead and merge this and file a separate issue re the problems I mentioned above!  At this point I would actually lean towards doing it that way rather than hold this up longer.

**Comment by byorgey** [comment]
A capability error reporting bug that I am likewise happy to split out into a separate issue after merging this, just want to write it down while it's fresh in my head:

- Start a classic mode game with `--cheat`
- Switch into creative mode and `create "branch predictor"`
- Type `build {move; if (1 > 2) {} {}}`
    - As expected we get an error saying no device provides the `compare` capability.
- Now type `build {move; turn left; if (1 > 2) {} {}}`
    - All we have done is add `turn left`; we would expect to get the same error.  Instead we get an error like `You do not have the devices required for: 'move; turn left; if (1 > 2) {noop} {noop}'  please obtain:   - branch predictor`
    - But we do have a branch predictor, so this error is bogus.  Not sure yet whether the capability checking or error reporting is to blame.

File: src/Swarm/Game/Exception.hs
Comment (xsebek) [reply src/Swarm/Game/Exception.hs]: I guess the Elden ring is now popular, so blasphemy it is:
```Haskell
incapableError (S.singleton CGod) (TConst As)
-- Thee shalt not utter such blasphemy:
--   'as'
--   If't be true thee wanteth to playeth god, then tryeth Creative game.
```

**Comment by xsebek** [comment]
@byorgey I think your first bug is present on main too, the Drill capability is not reported by `requiredCaps`.
The difference is, that I call that a `Fatal` error because `build` should have given it to the robot, while `main` drives the robots to rocks and there they stop without a drill.

See the debug logs I added:

### Main

Screenshot from 2022-06-14 17-01-12

### Iron
I was also missing standard devices, so I added them back:
Screenshot from 2022-06-14 16-48-55
Screenshot from 2022-06-14 17-19-14

---

I have no idea what the second bug is, but I hope it is not really my fault :smile:

**Pull request merged**
